### PR TITLE
[uikit] UICollectionView/UITableView Drag & Drop (Xcode 9 b1,2,3 & 4)

### PR DESCRIPTION
**Pull request by VincentDondain**

### CODE REVIEW

**Comment by VincentDondain** [comment]
Couple of specify points I'd like feedback on:

- `UISpringLoadedInteractionContext`, I believe it's right for it to have `[Model]` based on the description here: https://developer.apple.com/documentation/uikit/uispringloadedinteractioncontext.
- Same for `UISpringLoadedInteractionSupporting`: https://developer.apple.com/documentation/uikit/uispringloadedinteractionsupporting.
- API naming comments needed for `UICollectionViewDropCoordinator` and `UITableViewDropCoordinator`.
Especially for `DropItemToInsertedPlaceholder`, `DropItemToItem`, `DropItemIntoItem`, `DropItemToTarget`. There are many possibilities like actually "dropping" item in the name :P (`DropToItem`)...

File: src/uikit.cs [hunk 1]
```diff
@@ -3883,6 +3883,40 @@ interface UICollectionView : NSCoding {
...
+		[iOS (11,0)]
+		[Export ("dragInteractionEnabled")]
+		bool DragInteractionEnabled { get; set; }
+
+		[NoWatch, NoTV]
+		[iOS (11,0)]
+		[Export ("reorderingCadence", ArgumentSemantic.Assign)]
+		UICollectionViewReorderingCadence ReorderingCadence { get; set; }
+
+		[iOS (11,0)]
```
Comment (spouliot) [inline src/uikit.cs]: `[NoWatch, NoTV]` ?
or missing availability ?

File: src/uikit.cs [hunk 2]
```diff
@@ -11170,7 +11214,44 @@ interface UITableView : NSCoding {
...
+		[NullAllowed, Export ("dragDelegate", ArgumentSemantic.Weak)]
+		IUITableViewDragDelegate DragDelegate { get; set; }
+
+		[NoWatch, NoTV]
+		[iOS (11,0)]
+		[NullAllowed, Export ("dropDelegate", ArgumentSemantic.Weak)]
+		IUITableViewDropDelegate DropDelegate { get; set; }
+
+		[TV (11,0), iOS (11,0)]
```
Comment (spouliot) [inline src/uikit.cs]: `[NoWatch]` ?

File: src/uikit.cs [hunk 3]
```diff
@@ -11170,7 +11214,44 @@ interface UITableView : NSCoding {
...
+		[Export ("separatorInsetReference", ArgumentSemantic.Assign)]
+		UITableViewSeparatorInsetReference SeparatorInsetReference { get; set; }
+
+		[TV (11,0), iOS (11,0)]
```
Comment (spouliot) [inline src/uikit.cs]: `[NoWatch]` ?

File: src/uikit.cs [hunk 4]
```diff
@@ -11170,7 +11214,44 @@ interface UITableView : NSCoding {
...
+		[Async]
+		[Export ("performBatchUpdates:completion:")]
+		void PerformBatchUpdates ([NullAllowed] Action updates, [NullAllowed] Action<bool> completion);
 		
+		[TV (11,0), iOS (11,0)]
```
Comment (spouliot) [inline src/uikit.cs]: `[NoWatch]` ?

File: src/uikit.cs [hunk 5]
```diff
@@ -11696,6 +11787,23 @@ interface UITableViewDelegate {
 		[Export ("indexPathForPreferredFocusedViewInTableView:")]
 		[return: NullAllowed]
 		NSIndexPath GetIndexPathForPreferredFocusedView (UITableView tableView);
+
+		[NoTV]
```
Comment (spouliot) [inline src/uikit.cs]: `[NoWatch]` ?

File: src/uikit.cs [hunk 6]
```diff
@@ -11696,6 +11787,23 @@ interface UITableViewDelegate {
...
+		[iOS (11,0)]
+		[Export ("tableView:leadingSwipeActionsConfigurationForRowAtIndexPath:")]
+		[return: NullAllowed]
+		UISwipeActionsConfiguration GetLeadingSwipeActionsConfiguration (UITableView tableView, NSIndexPath indexPath);
+
+		[NoTV]
```
Comment (spouliot) [inline src/uikit.cs]: `[NoWatch]` ?

File: src/uikit.cs [hunk 7]
```diff
@@ -16407,6 +16529,449 @@ interface UITargetedDragPreview : NSCopying
...
+
+		[Export ("interactionEffect", ArgumentSemantic.Strong)]
+		IUISpringLoadedInteractionEffect InteractionEffect { get; }
+	}
+
+	interface IUISpringLoadedInteractionBehavior {}
+
+	[NoWatch, NoTV]
+	[iOS (11,0)]
+	[Protocol, Model]
```
Comment (dalexsoto) [inline src/uikit.cs]: Why [Model] here?

File: src/uikit.cs [hunk 8]
```diff
@@ -16407,6 +16529,449 @@ interface UITargetedDragPreview : NSCopying
...
+
+		[Export ("interactionDidFinish:")]
+		void InteractionDidFinish (UISpringLoadedInteraction interaction);
+	}
+
+	interface IUISpringLoadedInteractionEffect {}
+
+	[NoWatch, NoTV]
+	[iOS (11,0)]
+	[Protocol, Model]
```
Comment (dalexsoto) [inline src/uikit.cs]: Same, why [Model]?

File: src/uikit.cs [hunk 9]
```diff
@@ -16407,6 +16529,449 @@ interface UITargetedDragPreview : NSCopying
...
+		[Abstract]
+		[Export ("performUsingPresentationValues:")]
+		void PerformUsingPresentationValues (Action actionsToTranslate);
+	}
+
+	[NoWatch, NoTV]
+	[iOS (11,0)]
+	[BaseType (typeof(NSObject))]
+	[DisableDefaultCtor]
+	interface UISpringLoadedInteraction : IUIInteraction {
```
Comment (dalexsoto) [inline src/uikit.cs]: Why `IUIInteraction`? I think you meant `UIInteraction`

**Comment by VincentDondain** [comment]
Unrelated test failure: https://bugzilla.xamarin.com/show_bug.cgi?id=58394

**Comment by VincentDondain** [comment]
Please review the last commit which is the update to Xcode 9 beta 4.

File: src/uikit.cs [hunk 10]
```diff
@@ -16610,6 +16754,505 @@ interface UITargetedDragPreview : NSCopying
...
+		[Export ("proposal")]
+		UITableViewDropProposal Proposal { get; }
+
+		[Abstract]
+		[Export ("session")]
+		IUIDropSession Session { get; }
+
+		[Abstract]
+		[Export ("dropItem:toPlaceholder:")]
+		IUITableViewDropPlaceholderContext DropItemToPlaceholder (UIDragItem dragItem, UITableViewDropPlaceholder placeholder);
```
Comment (spouliot) [inline src/uikit.cs]: `DropItem` or even just `Drop` (like Swift)

File: src/uikit.cs [hunk 11]
```diff
@@ -16610,6 +16754,505 @@ interface UITargetedDragPreview : NSCopying
...
+
+		[Abstract]
+		[Export ("dropItem:toRowAtIndexPath:")]
+		IUIDragAnimating DropItemToRow (UIDragItem dragItem, NSIndexPath indexPath);
```
Comment (spouliot) [inline src/uikit.cs]: same

File: src/uikit.cs [hunk 12]
```diff
@@ -16610,6 +16754,505 @@ interface UITargetedDragPreview : NSCopying
...
+
+		[Abstract]
+		[Export ("dropItem:intoRowAtIndexPath:rect:")]
+		IUIDragAnimating DropItemIntoRow (UIDragItem dragItem, NSIndexPath indexPath, CGRect rect);
```
Comment (spouliot) [inline src/uikit.cs]: same

File: src/uikit.cs [hunk 13]
```diff
@@ -16610,6 +16754,505 @@ interface UITargetedDragPreview : NSCopying
...
+
+		[Abstract]
+		[Export ("dropItem:toTarget:")]
+		IUIDragAnimating DropItemToTarget (UIDragItem dragItem, UIDragPreviewTarget target);
```
Comment (spouliot) [inline src/uikit.cs]: same

File: src/uikit.cs [hunk 14]
```diff
@@ -16610,6 +16754,505 @@ interface UITargetedDragPreview : NSCopying
...
+	[NoWatch, NoTV, iOS (11,0)]
+	[Protocol]
+	interface UITableViewDropPlaceholderContext : UIDragAnimating {
+		[Abstract]
+		[Export ("dragItem")]
+		UIDragItem DragItem { get; }
+
+		[Abstract]
+		[Export ("commitInsertionWithDataSourceUpdates:")]
+		bool CommitInsertionWithDataSourceUpdates (Action<NSIndexPath> dataSourceUpdates);
```
Comment (spouliot) [inline src/uikit.cs]: ` CommitInsertion`

File: src/uikit.cs [hunk 15]
```diff
@@ -16610,6 +16754,505 @@ interface UITargetedDragPreview : NSCopying
...
+	[iOS (11,0)]
+	delegate void UIContextualActionCompletionHandler (bool finished);
+
+	[NoWatch, NoTV]
+	[iOS (11,0)]
+	[BaseType (typeof(NSObject))]
+	interface UIContextualAction {
+		[Static]
+		[Export ("contextualActionWithStyle:title:handler:")]
+		UIContextualAction FromContextualActionStyle (UIContextualActionStyle style, [NullAllowed] string title, UIContextualActionHandler handler);
```
Comment (spouliot) [inline src/uikit.cs]: this static method makes me think `init` should not be allowed -> `[DisableDefaultCtor]`

File: src/uikit.cs [hunk 16]
```diff
@@ -16610,6 +16754,505 @@ interface UITargetedDragPreview : NSCopying
...
+		UIImage Image { get; set; }
+	}
+
+	[NoWatch, NoTV]
+	[iOS (11,0)]
+	[BaseType (typeof(NSObject))]
+	interface UISwipeActionsConfiguration {
+		[Static]
+		[Export ("configurationWithActions:")]
+		UISwipeActionsConfiguration FromActions (UIContextualAction[] actions);
```
Comment (spouliot) [inline src/uikit.cs]: same -> `[DisableDefaultCtor]`

File: src/uikit.cs (hunk 10 above)
Comment (VincentDondain) [inline src/uikit.cs]: @sebastien mmh it’s kinda confusing in C# as I don’t think we should name the 2nd argument `toPlaceholder` right.

We'd get:

``` C#
IUICollectionViewDropPlaceholderContext Drop (UIDragItem dragItem, UICollectionViewDropPlaceholder placeholder);

IUIDragAnimating Drop (UIDragItem dragItem, NSIndexPath indexPath);

IUIDragAnimating Drop (UIDragItem dragItem, NSIndexPath indexPath, CGRect rect);

IUIDragAnimating Drop (UIDragItem dragItem, UIDragPreviewTarget target);
```

We're loosing context here when calling those methods compared to Swift where the `to` or `toItemAt` stays when calling the method:

``` Swift
drop(myDragItem, to: myPlaceholder)
drop(myDragItem, toItemAt: myIndexPath)
drop(myDragItem, intoItemAt: myIndexPath, rect: myRect)
drop(myDragItem, to: myTarget)
```

We'd have 4 `Drop` methods and I don't think the arguments types alone give enough context.

I think we can go `DropToPlaceholder`, `DropToItem`, etc, etc... though.

Do you agree?

Comment (spouliot) [inline src/uikit.cs]: good point, let's use longer, clearer names

**Pull request merged**